A29-xds-tls-security.md

```diff
@@ -0,0 +1,509 @@
+watcher.
+
+The **`CertificateProvider`** plugin also has an associated factory
+(**`CertificateProviderFactory`**) that is used to instantiate the plugin. The
+factory is identified by a unique name e.g. `"file_watcher"` that is also the identity of the
+plugin. The factory is used to instantiate the plugin after validating the received
+configuration. The framework uses deduping and reference counting to ensure that a plugin
+of a given kind and configuration is instantiated only once.
+
+A plugin typically creates a new thread and a timer to periodically query and 
+send the certificate (and the key) to all its watchers. **`FileWatcherCertificateProvider`**
+(aka `file_watcher` described [above][FILE-WATCHER-LINK]) is an implementation of
+the **`CertificateProvider`**. This plugin monitors configured file-paths in the file
+system and reads those files on updates to get the latest certificates and keys and
+provide those to the watchers after converting them to the canonical format. For example,
+Java uses `java.security.PrivateKey` and `java.security.cert.X509Certificate` for a
+private key and a certificate respectively.
+
+## Rationale
+
+gRPC has been steadily adding support for various xDS features such as load balancing and
+advanced traffic management. xDS-based security was next in the roadmap and this proposal
+addresses that requirement. gRPC makes a proxyless service mesh (PSM) possible. A secure
+proxyless service mesh requires the security features described in this proposal.
+
+The Certificate Provider Plugin Framework provides a generic alternative to the SDS
+server/agent based solution and eliminates the dependency on the SDS protocol. An SDS
+protocol based client can still be added just as another plugin implementation in the
+framework. The plugin framework allows different environments to have different
+dependencies because the xDS client environment is abstracted out via the plugin
+framework and the control plane can be agnostic to the specific client environment.
+The framework has also made the certificate provider functionality extensible
+and pluggable which enables support for new certificate providers without requiring
+changes to the xDS protocol or the control plane. These advantages plus the extensions
+made to the xDS protocol are providing an impetus to the Envoy community so that Envoy
+will hopefully have a similar framework in the near future to support pluggable and
+extensible certificate providers.
+
+In any case the addition of xDS based security has brought gRPC xDS support closer to
+Envoy in terms of capabilities. This enables various use-cases such as migration to
+proxyless and interop or co-existence with Envoy.
+
+## Implementation Status
+
+Java had an early prototype implementation of xDS-based security that used SDS similar to Envoy. It was
+then modified and refined to match this spec. Other languages followed soon after to complete
```

Comment:
1. Should we clarify exact versions when this happened? 
2. Or should we just drop this part and describe current state of things instead?

Reply:
Describing the current state includes how the implementation came about and the reason for including this context. I don't know the value of adding versions - also there are no "exact" versions since this happened incrementally